source/slang/slang-ir-transform-params-to-constref.cpp

```diff
@@ -0,0 +1,325 @@
+                    switch (user->getOp())
+                    {
+                    case kIROp_FieldExtract:
+                        {
+                            // Transform the IRFieldExtract into a IRFieldAddress
+                            auto fieldExtract = as<IRFieldExtract>(use->getUser());
+                            builder.setInsertBefore(fieldExtract);
+                            auto fieldAddr = builder.emitFieldAddress(
+                                fieldExtract->getBase(),
+                                fieldExtract->getField());
+                            auto loadInst = builder.emitLoad(fieldAddr);
+                            fieldExtract->replaceUsesWith(loadInst);
+                            fieldExtract->removeAndDeallocate();
+                            break;
+                        }
+                    case kIROp_GetElement:
+                        {
+                            // Transform the IRGetElement into a IRGetElementPtr
+                            auto getElement = as<IRGetElement>(use->getUser());
+
+                            builder.setInsertBefore(getElement);
+                            auto elemAddr = builder.emitElementAddress(
+                                getElement->getBase(),
+                                getElement->getIndex());
+                            auto loadInst = builder.emitLoad(elemAddr);
+                            getElement->replaceUsesWith(loadInst);
+                            getElement->removeAndDeallocate();
+                            break;
+                        }
+                    default:
+                        {
+                            // Insert a load before the user and replace the user with the load
+                            builder.setInsertBefore(user);
+                            auto loadInst = builder.emitLoad(param);
+                            use->set(loadInst);
+                            break;
+                        }
+                    }
+                });
+        }
+    }
+
+    // Check if an instruction represents an addressable value
+    IRInst* getAddressable(IRInst* inst)
+    {
+        if (!inst)
+            return nullptr;
+
+        switch (inst->getOp())
+        {
+        case kIROp_Var:
+        case kIROp_GlobalVar:
+        case kIROp_FieldAddress:
+        case kIROp_GetElementPtr:
+            return inst;
+        case kIROp_GlobalParam:
+            return builder.emitLoad(inst);
+        case kIROp_FieldExtract:
+        case kIROp_GetElement:
+        case kIROp_Param:
+            return builder.emitGetAddress(builder.getPtrType(inst->getDataType()), inst);
+        case kIROp_Load:
+            // Check if the load is from an addressable source.
+            // Track the target operand since if we have a
+            // `load(x)`, we want to pass x as our result
+            // since `x` would be the address of `load(x)`.
+            return as<IRLoad>(inst)->getPtr();
```

Comment:
I don't think this is correct. This should be left to redundant load store removal because we need to make sure there are no side-effects in between the load and the call.